# Day 12, part 1

The pots are numbered, with 0 in front of you. To the left, the pots are numbered -1, -2, -3, and so on; to the right, 1, 2, 3.... Your puzzle input contains a list of pots from 0 to the right and whether they do (#) or do not (.) currently contain a plant, the initial state. (No other pots currently contain plants.) For example, an initial state of #..##.... indicates that pots 0, 3, and 4 currently contain plants.

Your puzzle input also contains some notes you find on a nearby table: someone has been trying to figure out how these plants spread to nearby pots. Based on the notes, for each generation of plants, a given pot has or does not have a plant based on whether that pot (and the two pots on either side of it) had a plant in the last generation. These are written as LLCRR => N, where L are pots to the left, C is the current pot being considered, R are the pots to the right, and N is whether the current pot will have a plant in the next generation.

## Q: After 20 generations, what is the sum of the numbers of all pots which contain a plant?

correct answer: 2571

In [7]:
########### functions ##############
# add two dots at beginning and at end
def add_start_end(state: dict):
    min_ind = min(state)
    max_ind = max(state)
    for i in range(min_ind-3, min_ind):
        state[i] = '.'
    for i in range(max_ind+1, max_ind+4):
        state[i] = '.'
        
# get a 5 char string from dictionary (-2 to +2 values)
def get_config(i: int, state: dict) -> str:
    return ''.join(state.get(j, '.') for j in range(i-2, i+3))
    

# calculate next generation
def calc_next(state: dict) -> dict:
    next_gen = {}
    # add empty pots to left and right end
    add_start_end(state)
    for p in range(min(state), max(state)+1):
        test = get_config(p, state)
        val = instructions.get(test, '.')
        #print('%s -> %s' % (test, val))
        next_gen[p] = val
    return next_gen

# print out the generation map
def print_gens(gen_list):
    header_str = ''.join(' ' if i % 10 else 'x' for i in range(-10, 150))
    print('{0:3d} {1}'.format(0, header_str))
    for i in range(len(gen_list)):
        state_str = ''.join(gen_list[i].get(j, '.') for j in range(-10, 150))
        print('{0:3d} {1}'.format(i, state_str))

########## main #############
# number of generations
gens = 200
# create list of dictionaries - each dictionary is one generation
gen = [dict() for _ in range(gens+1)]

# define dict to count
pot_val_dict = {'.': 0, '#': 1}

# read initial state and create generation 0 
my_file = open(r'D:\Python\Advent\12.1\input.txt', 'r')
initial_state = list(my_file.readline().strip()[15:])
gen[0] = dict(enumerate(initial_state))

# discard empty line
my_file.readline()

# read in instructions into dictionary
instructions = {}
for l in my_file.readlines():
    inst = l[:5]
    res = l[-2]
    instructions[inst] = res

# calculate next 20 generations
for i in range(1, gens+1):
    gen[i] = calc_next(gen[i-1])

# calculate value for 20th generation
gen20 = gen[gens]
part1 = sum(i * pot_val_dict[gen20[i]] for i in range(min(gen20), max(gen20)+1))
    
print(part1)


13055


In [8]:
part2_val = [0 for i in range(200)]
for i in range(1, 200):
    gen_x = gen[i]
    part2_val[i] = sum(i * pot_val_dict[gen_x[i]] for i in range(min(gen_x), max(gen_x)+1))
    print('{0:3d} {1} delta: {2}'.format(i, part2_val[i], part2_val[i]-part2_val[i-1]))


  1 2823 delta: 2823
  2 2860 delta: 37
  3 2740 delta: -120
  4 2669 delta: -71
  5 2886 delta: 217
  6 3015 delta: 129
  7 3003 delta: -12
  8 3011 delta: 8
  9 3011 delta: 0
 10 2758 delta: -253
 11 3000 delta: 242
 12 3348 delta: 348
 13 3283 delta: -65
 14 2924 delta: -359
 15 2948 delta: 24
 16 2857 delta: -91
 17 3038 delta: 181
 18 3289 delta: 251
 19 3072 delta: -217
 20 2571 delta: -501
 21 2979 delta: 408
 22 3040 delta: 61
 23 3108 delta: 68
 24 3303 delta: 195
 25 3491 delta: 188
 26 3395 delta: -96
 27 3192 delta: -203
 28 3368 delta: 176
 29 3431 delta: 63
 30 3098 delta: -333
 31 3326 delta: 228
 32 3250 delta: -76
 33 3471 delta: 221
 34 3518 delta: 47
 35 3385 delta: -133
 36 3407 delta: 22
 37 3702 delta: 295
 38 3381 delta: -321
 39 3656 delta: 275
 40 3615 delta: -41
 41 3671 delta: 56
 42 3679 delta: 8
 43 3735 delta: 56
 44 3790 delta: 55
 45 3792 delta: 2
 46 4029 delta: 237
 47 4051 delta: 22
 48 3999 delta: -52
 49 4126 delta: 127
 50 4103 delta: -23
 51 4098 

In [95]:
# the score stabilizes to 62 at gen 100
# to get the 50000000000 value, take the value for 100 and multiply (5bn - 100) * 62
c = 50000000000 - 100
part2 = 6855 + c * 62
print(part2)

3100000000655


# part 2

You realize that 20 generations aren't enough. After all, these plants will need to last another 1500 years to even reach your timeline, not to mention your future.

After fifty billion (50000000000) generations, what is the sum of the numbers of all pots which contain a plant?

correct answer: 3100000000655